source/websocket_encoder.c

```diff
@@ -0,0 +1,347 @@
+    /* Fill tmp buffer with extended-length in network byte order */
+    uint8_t network_bytes_array[8];
+    struct aws_byte_buf network_bytes_buf =
+        aws_byte_buf_from_empty_array(network_bytes_array, sizeof(network_bytes_array));
+    if (encoder->frame.payload_length <= AWS_WEBSOCKET_2BYTE_EXTENDED_LENGTH_MAX_VALUE) {
+        aws_byte_buf_write_be16(&network_bytes_buf, (uint16_t)encoder->frame.payload_length);
+    } else {
+        aws_byte_buf_write_be64(&network_bytes_buf, encoder->frame.payload_length);
+    }
+
+    /* Use cursor to iterate over tmp buffer */
+    struct aws_byte_cursor network_bytes_cursor = aws_byte_cursor_from_buf(&network_bytes_buf);
+
+    /* Advance cursor if some bytes already written */
+    aws_byte_cursor_advance(&network_bytes_cursor, (size_t)encoder->state_bytes_processed);
+
+    /* Shorten cursor if it won't all fit in out_buf */
+    bool all_data_written = true;
+    size_t space_available = out_buf->capacity - out_buf->len;
+    if (network_bytes_cursor.len > space_available) {
+        network_bytes_cursor.len = space_available;
+        all_data_written = false;
+    }
+
+    aws_byte_buf_write_from_whole_cursor(out_buf, network_bytes_cursor);
+    encoder->state_bytes_processed += network_bytes_cursor.len;
+
+    /* If all bytes written, advance to next state */
+    if (all_data_written) {
+        encoder->state = AWS_WEBSOCKET_ENCODER_STATE_MASKING_KEY_CHECK;
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+/* MASKING_KEY_CHECK: Outputs no data. Gets things ready for (or decides to skip) the STATE_MASKING_KEY */
+static int s_state_masking_key_check(struct aws_websocket_encoder *encoder, struct aws_byte_buf *out_buf) {
+    (void)out_buf;
+
+    if (encoder->frame.masked) {
+        encoder->state_bytes_processed = 0;
+        encoder->state = AWS_WEBSOCKET_ENCODER_STATE_MASKING_KEY;
+    } else {
+        encoder->state = AWS_WEBSOCKET_ENCODER_STATE_PAYLOAD_CHECK;
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+/* MASKING_KEY: Output masking-key (state skipped if no masking key). */
+static int s_state_masking_key(struct aws_websocket_encoder *encoder, struct aws_byte_buf *out_buf) {
+    /* Prepare cursor to iterate over masking-key bytes */
+    struct aws_byte_cursor cursor =
+        aws_byte_cursor_from_array(encoder->frame.masking_key, sizeof(encoder->frame.masking_key));
+
+    /* Advance cursor if some bytes already written */
+    aws_byte_cursor_advance(&cursor, (size_t)encoder->state_bytes_processed);
+
+    /* Shorten cursor if it won't all fit in out_buf */
+    bool all_data_written = true;
+    size_t space_available = out_buf->capacity - out_buf->len;
+    if (cursor.len > space_available) {
+        cursor.len = space_available;
+        all_data_written = false;
+    }
+
+    aws_byte_buf_write_from_whole_cursor(out_buf, cursor);
+    encoder->state_bytes_processed += cursor.len;
```

Comment:
This took me a few minutes to work out, maybe a comment would help, because advance also shrinks the len. Offhand, it looks like you're adding too much here.

Reply:
Added comments.
We should probably be using byte_cursor advance more often. The class is supposed to be like a "cursor" that moves forward as you use it. It's got some weirdness I asked about in the c-chat the other day.